Cleaning up the transformation code

jennej opened the pull request:
_New and _AffectReturnValuePassedIn have been almost entirely eradicated from the codebase

I would like to go a bit further with the PicturePlaneUtils functions, but this was getting massive, and it seems like some of it might happen naturally if I dive deeper into the pose code, so I left things as they were.

jennej committed:
passing time

jennej committed:
get rid of the special case camera versions of these functions

jennej committed:
removing some unused code

jennej committed:
all AWT transforms are "new" transforms

the not "new" ones were never being used and it was hard to read

also, remove unused functions

jennej committed:
added the ability to just get a bounding box as vectors

we were creating more objects than we needed, it probably won't matter much

jennej committed:
All component transforms are the same

and the same is New.  muck with your own variables, don't expect transformations to do it for you any more.

jennej committed:
simplify TransformationUtilities

There are now three functions offered by transformationUtilities

transformToAbsolute
transformFromAbsolute
transformTo

and we offer them for Vector4, Vector3, and Point3.
(not all functions are used for all types, but it is not unreasonable to think they might be)

all of them return a brand new object, created bespoke for the calling function.

jennej committed:
now with comments

the bar is low

jennej committed:
removing cleverness

* less static variables/syncronized blocks
* just make the variables you need where you need them

jennej committed:
removing another static variable

jennej committed:
make functions private

_AffectReturnValuePassedIn is now our little secret

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/render/PicturePlaneUtils.java:
```diff
@@ -157,185 +155,38 @@ private static Vector4 transformFromCameraToViewport_AffectReturnValuePassedIn(V
-    rv.set(xyzw);
+  public static Vector4 transformFromViewportToCamera(Vector4 xyzw, RenderTarget renderTarget, AbstractCamera sgCamera) {
+    Vector4 rv = new Vector4(xyzw);
     return transformFromViewportToCamera_AffectReturnValuePassedIn(rv, renderTarget, sgCamera);
```
This, and others here, still have the odd suffixes. They are all private, so I can see where you drew the line.
But this one, and some of the others, have only one caller and could be inlined and removed.
That can be for a next round of fixes.

jennej replied:
Yeah, this is the stuff I'm hoping I can sort out/possibly eliminate with the pose code work. I have them gone in a stash.

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/scenegraph/AbstractCamera.java:
```diff
@@ -43,41 +43,13 @@
  * @author Dennis Cosgrove
  */
 public abstract class AbstractCamera extends Leaf {
-  public Point transformToAWT(Point rv, Vector4 xyzw, RenderTarget renderTarget) {
```
Having these on camera (or perhaps RenderTarget) makes some sense.
Having the logic bounce up through Component and out to static helper methods, less so, so I can see why you took them out.
If you are looking for a non-static home for some of the functions, these might return - with implementations not passthroughs.

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/scenegraph/util/TransformationUtilities.java:
The remaining `public static` methods are called from Component and perhaps should be methods there.
This might happen in this PR, another PR, or not at all.

jennej replied:
oh, i was in too deep to see this, thank you.

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/render/PicturePlaneUtils.java:
```diff
@@ -55,22 +53,22 @@
-    rv.z = z;
-    rv.w = 1.0;
-    return rv;
+  private static Vector4 transformFromAWTToViewport(Point p, double z, Rectangle actualViewport) {
```
This has no callers. Do we need it?

dabeshouse reviewed (approved):
There are more changes that could be made.
They could be put in this PR, but don't need to be.

jennej committed:
removing utilities, not utility

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/scenegraph/Component.java:
```diff
@@ -238,37 +237,107 @@ protected void fireHierarchyChanged() {
 
+  private boolean isAbsolute() {
+    //todo: call isSceneOf()?
+    if ((this instanceof Scene) || ((ReferenceFrame) this == AsSeenBy.SCENE)) {
```
Now that this is on Component the ReferenceFrame is more obviously not the case, so it simplifies to `instanceOf`, or just false with an overriding true in Scene.

jennej committed:
re-precate getActualViewportAsAwtRectangle

we can deprecate it once we have a plan. or better yet, we can just replace it if there's something better

also, get rid of another unnecessary static variable

jennej committed:
render target/camera cleanups

removing static variables

getting rid of a bunch of  rvs in the renderTarget/camera worlds so that they are at least moderately consistent with themselves

jennej committed:
simplify isAbsolute

PR feedback

jennej committed:
move PicturePlaneUtils into Render Target

there are absolutely more cleanups that can happen on these functions, but i'm holding off til i figure out what is going on in PicturePlaneInteractions.

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/scenegraph/Component.java:
```diff
@@ -258,114 +234,93 @@ protected void fireHierarchyChanged() {
-    return from.transformTo_New(xyz, this);
+  // Vector4
+  public Vector4 transformToAbsolute(Vector4 xyz) {
+    Vector4 rv = new Vector4(xyz);
```
Could be inside conditional. No need to make a new one if it is the same.

dabeshouse commented on core/scenegraph/src/main/java/edu/cmu/cs/dennisc/scenegraph/Component.java:
```diff
@@ -258,114 +234,93 @@ protected void fireHierarchyChanged() {
-  public Vector4 transformTo_AffectReturnValuePassedIn(Vector4 xyzw, Component to) {
-    return TransformationUtilities.transformTo_AffectReturnValuePassedIn(xyzw, this, to);
+  public Vector4 transformFromAbsolute(Vector4 xyz) {
+    Vector4 rv = new Vector4(xyz);
```
As above, this could be inside the conditional. And a few more below…

dabeshouse reviewed (approved):
lgtm
If you are looking for more, I pointed out minor Component tweaks, or they can be in the next round, or not.
